tests/storages_tests/rdb_tests/test_storage.py

```diff
@@ -364,3 +365,29 @@ def test_invalid_heartbeat_interval_and_grace_period() -> None:
 
     with pytest.raises(ValueError):
         _ = RDBStorage("sqlite:///:memory:", grace_period=-1)
+
+
+def test_failed_trial_callback() -> None:
+    heartbeat_interval = 1
+    grace_period = 2
+
+    def failed_trial_callback(study: Study, trial: FrozenTrial) -> None:
```

Comment:
How about checking that this function is actually called? This test may pass when the callback is not executed.